util/network/cniprovider/cni.go

```diff
@@ -67,19 +82,153 @@ func (c *cniProvider) initNetwork() error {
+	// LIFO: Ordered least recently used to most recently used
+	available []*cniNS
+	closed    bool
+}
+
+func (pool *cniPool) close() {
+	bklog.L.Debugf("cleaning up cni pool")
+
+	pool.mu.Lock()
+	pool.closed = true
+	defer pool.mu.Unlock()
+	for len(pool.available) > 0 {
+		_ = pool.available[0].release()
+		pool.available = pool.available[1:]
+		pool.actualSize--
+	}
+}
+
+func (pool *cniPool) fillPool(ctx context.Context) {
+	for {
+		pool.mu.Lock()
+		if pool.closed {
+			pool.mu.Unlock()
+			return
+		}
+		actualSize := pool.actualSize
+		pool.mu.Unlock()
+		if actualSize >= pool.targetSize {
+			return
+		}
+		ns, err := pool.getNew(ctx)
+		if err != nil {
+			bklog.G(ctx).Errorf("failed to create new network namespace while prefilling pool: %s", err)
+			return
+		}
+		pool.put(ns)
+	}
+}
+
+func (pool *cniPool) get(ctx context.Context) (*cniNS, error) {
+	pool.mu.Lock()
+	if len(pool.available) > 0 {
+		ns := pool.available[len(pool.available)-1]
+		pool.available = pool.available[:len(pool.available)-1]
+		pool.mu.Unlock()
+		trace.SpanFromContext(ctx).AddEvent("returning network namespace from pool")
+		bklog.G(ctx).Debugf("returning network namespace %s from pool", ns.id)
+		return ns, nil
+	}
+	pool.mu.Unlock()
+
+	return pool.getNew(ctx)
+}
+
+func (pool *cniPool) getNew(ctx context.Context) (*cniNS, error) {
+	ns, err := pool.provider.newNS(ctx, "")
+	if err != nil {
+		return nil, err
+	}
+	ns.pool = pool
+
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+	if pool.closed {
+		return nil, errors.New("cni pool is closed")
+	}
+	pool.actualSize++
+	return ns, nil
+}
+
+func (pool *cniPool) put(ns *cniNS) {
+	putTime := time.Now()
+	ns.lastUsed = putTime
+
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+	if pool.closed {
+		_ = ns.release()
+		return
+	}
+	pool.available = append(pool.available, ns)
+	actualSize := pool.actualSize
+
+	if actualSize > pool.targetSize {
+		// We have more network namespaces than our target number, so
+		// schedule a shrinking pass.
+		time.AfterFunc(aboveTargetGracePeriod, pool.cleanupToTargetSize)
+	}
+}
+
+func (pool *cniPool) cleanupToTargetSize() {
+	var toRelease []*cniNS
+	defer func() {
+		for _, poolNS := range toRelease {
+			_ = poolNS.release()
+		}
+	}()
+
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+	for pool.actualSize > pool.targetSize &&
+		len(pool.available) > 0 &&
+		time.Since(pool.available[0].lastUsed) >= aboveTargetGracePeriod {
+		bklog.L.Debugf("releasing network namespace %s since it was last used at %s", pool.available[0].id, pool.available[0].lastUsed)
+		toRelease = append(toRelease, pool.available[0])
+		pool.available = pool.available[1:]
+		pool.actualSize--
+	}
+}
+
+func (c *cniProvider) New(ctx context.Context, hostname string) (network.Namespace, error) {
+	// We can't use the pool for namespaces that need a custom hostname.
+	// We also avoid using it on windows because we don't have a cleanup
+	// mechanism for Windows yet.
+	if hostname == "" || runtime.GOOS == "windows" {
```

Comment:
isn't this supposed to be `&& runtime.GOOS != "windows"` . If not then I think the comment is confusing.